File: backend/internal/models/bookings.go
// backend/internal/models/bookings.go
package models

import "time"

type Booking struct {
	ID              string    `json:"id" db:"id"`
	UserID          string    `json:"user_id" db:"user_id"`
	BusID           string    `json:"bus_id" db:"bus_id"`
	RouteID         string    `json:"route_id" db:"route_id"`
	BoardingStopID  string    `json:"boarding_stop_id" db:"bus_stop_id"`
	BoardingStopName string    `json:"boarding_stop_name" db:"boarding_stop_name"`
	AlightingStopID string    `json:"alighting_stop_id" db:"alighting_stop_id"`
	AlightingStopName string    `json:"alighting_stop_name" db:"alighting_stop_name"`
	Seats           SeatMap   `json:"seats" db:"seats"`
	Fare            float64   `json:"fare" db:"fare"`
	BookingTime     time.Time `json:"booking_time" db:"booking_time"`
	Status          string    `json:"status" db:"status"`
	CreatedAt       time.Time `json:"created_at" db:"created_at"`
	ExpiresAt       time.Time `json:"expires_at" db:"expires_at"`
	BoardedAt       time.Time `json:"boarded_at" db:"boarded_at"`
}

type SeatMap struct {
	SeatNumbers []string `json:"seat_numbers"`
	Count       int      `json:"count"`
}

File: frontend/lib/api/bookings.ts
import { api, ApiResponse } from '../api';

export interface CreateBookingRequest {
  bus_id: string;
  seats: {
    seat_numbers: string[];
    count: number;
  };
  payment_method: 'mpesa' | 'airtel_money' | 'cash' | 'bus_pass';
}

export interface Booking {
  id: string;
  user_id: string;
  bus_id: string;
  route_id: string;
  seats: {
    seat_numbers: string[];
    count: number;
  };
  fare: number;
  status: 'pending' | 'confirmed' | 'cancelled' | 'completed';
  created_at: string;
  expires_at: string;
  boarded_at?: string;
}

export const bookingsApi = {
  // Create a new booking
  createBooking: async (data: CreateBookingRequest): Promise<ApiResponse<Booking>> => {
    const response = await api.post<Booking>('/bookings', data);
    return {
      data: response.data,
      status: response.status,
    };
  },
  
  // Get a booking by ID
  getBooking: async (bookingId: string): Promise<ApiResponse<Booking>> => {
    const response = await api.get<Booking>(`/bookings/${bookingId}`);
    return {
      data: response.data,
      status: response.status,
    };
  },
  
  // Cancel a booking
  cancelBooking: async (bookingId: string): Promise<ApiResponse<{ message: string }>> => {
    const response = await api.post<{ message: string }>(`/bookings/${bookingId}/cancel`);
    return {
      data: response.data,
      status: response.status,
    };
  },
  
  // Get user's bookings
  getUserBookings: async (): Promise<ApiResponse<Booking[]>> => {
    const response = await api.get<Booking[]>('/me/bookings');
    return {
      data: response.data,
      status: response.status,
    };
  },
}; 